File: 0120.三角形最小路径和/0120-三角形最小路径和.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int ans = INT_MAX;
        int n = triangle.size();
        if( n == 0 ) {
            return 0;
        }

        for( int i=1; i<n; i++ ) {
            for( int j=0; j<=i; j++ ) {
                if( j == 0 ) {
                    triangle[i][j] += triangle[i-1][j] ;
                } else if( j == i ){
                    triangle[i][j] += triangle[i-1][j-1];
                } else {
                    triangle[i][j] += min(triangle[i-1][j], triangle[i-1][j-1]);
                }
            }
        }

        for( int j=0; j<n; j++ ) {
            ans = min( ans, triangle[n-1][j] );
        }

        return ans;
    }
};

File: 0120.三角形最小路径和/0120-三角形最小路径和.go
func minimumTotal(triangle [][]int) int {
    n := len(triangle)
    
    for i:=1; i<n; i++ {
        for j:=0; j<=i; j++ {
            if j == 0 {
                triangle[i][j] += triangle[i-1][j]
            } else if j == i {
                triangle[i][j] += triangle[i-1][j-1]
            } else {
                triangle[i][j] += min(triangle[i-1][j-1], triangle[i-1][j])
            }
        }
    }

    minTotal := triangle[n-1][0]
    for j:=1; j<n; j++ {
        minTotal = min(triangle[n-1][j], minTotal)
    }

    return minTotal
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}